include/geos/geom/util/GeometryFixer.h

```diff
@@ -0,0 +1,142 @@
+/**********************************************************************
+ *
+ * GEOS - Geometry Engine Open Source
+ * http://geos.osgeo.org
+ *
+ * Copyright (C) 2021 Paul Ramsey <pramsey@cleverelephant.ca>
+ *
+ * This is free software; you can redistribute and/or modify it under
+ * the terms of the GNU Lesser General Public Licence as published
+ * by the Free Software Foundation.
+ * See the COPYING file for more information.
+ *
+ **********************************************************************/
+
+#pragma once
+
+#include <memory>
+#include <vector>
+
+#include <geos/geom/Geometry.h>
+#include <geos/export.h>
+
+// Forward declarations
+namespace geos {
+namespace geom {
+class GeometryFactory;
+class Point;
+class Polygon;
+class LineString;
+class LinearRing;
+class MultiPoint;
+class MultiPolygon;
+class MultiLineString;
+class GeometryCollection;
+}
+}
+
+namespace geos {
+namespace geom { // geos.geom
+namespace util { // geos.geom.util
+
+/**
+ * Fixes a geometry to be a valid geometry, while preserving as much as
+ * possible of the shape and location of the input.
+ * Validity is determined according to {@link Geometry#isValid()}.
+ *
+ * Input geometries are always processed, so even valid inputs may
+ * have some minor alterations.  The output is always a new geometry object.
+ *
+ * <h2>Semantic Rules</h2>
+ *
+ * * Vertices with non-finite X or Y ordinates are removed (as per Coordinate::isValid() ).
+ *  * Repeated points are reduced to a single point
+ *  * Empty atomic geometries are valid and are returned unchanged
+ *  * Empty elements are removed from collections
+ *  * <code>Point</code>: keep valid coordinate, or EMPTY
+ *  * <code>LineString</code>: fix coordinate list
+ *  * <code>LinearRing</code>: fix coordinate list, return as valid ring or else <code>LineString</code>
+ *  * <code>Polygon</code>: transform into a valid polygon, preserving as much of the extent and vertices as possible
+ *  * <code>MultiPolygon</code>: fix each polygon, then ensure result is non-overlapping (via union)
+ *  * <code>GeometryCollection</code>: fix each element
+ *  * Collapsed lines and polygons are handled as follows,
+ *
+ * depending on the <code>keepCollapsed</code> setting:
+ *
+ *  * <code>false</code>: (default) collapses are converted to empty geometries
+ *  * <code>true</code>: collapses are converted to a valid geometry of lower dimension
+ *
+ * @author Martin Davis
+*/
+class GEOS_DLL GeometryFixer {
+
+private:
+
+    const geom::Geometry* geom;
+    const geom::GeometryFactory* factory;
+    bool isKeepCollapsed; // false
+
+public:
+
+    GeometryFixer(const geom::Geometry* p_geom)
```

Comment:
a const reference might be slightly better than a const pointer, as it is then the responsibility of the caller to make sure they don't reference a null pointer. But the current approach is probably more in line with the rest of the code base

Reply:
Yeah, it kind of matches current patterns. I'm inclined to leave as is.